tests/ct_commands.c

```diff
@@ -72,6 +72,135 @@ void test_mget(redisClusterContext *cc) {
+    reply = (redisReply *)redisClusterCommand(
+        cc, "HDEL myhash field1 field2 field3");
+    CHECK_REPLY_INT(cc, reply, 1); // field3 deleted
+    freeReplyObject(reply);
+
+    // Make sure field3 is deleted now
+    reply = (redisReply *)redisClusterCommand(cc, "HEXISTS myhash field3");
+    CHECK_REPLY_INT(cc, reply, 0); // no field
+    freeReplyObject(reply);
+
+    // As of Redis 4.0.0, HSET is variadic i.e. multiple field/value,
+    // but this is currently not supported.
+    reply = (redisReply *)redisClusterCommand(
+        cc, "HSET myhash field1 hsetvalue1 field2 hsetvalue2");
+    assert(reply == NULL);
+}
+
+// Command layout:
+// eval <script> <no of keys> <keys..> <args..>
+void test_eval(redisClusterContext *cc) {
+    redisReply *reply;
+
+    // Single key
+    reply = (redisReply *)redisClusterCommand(
+        cc, "eval %s 1 %s", "return redis.call('set',KEYS[1],'bar')", "foo");
+    CHECK_REPLY_OK(cc, reply);
+    freeReplyObject(reply);
+
+    // Single key, string response
+    reply = (redisReply *)redisClusterCommand(cc, "eval %s 1 %s",
+                                              "return KEYS[1]", "key1");
+    CHECK_REPLY_STR(cc, reply, "key1");
+    freeReplyObject(reply);
+
+    // Single key with single argument
+    reply = (redisReply *)redisClusterCommand(
+        cc, "eval %s 1 %s %s", "return {KEYS[1],ARGV[1]}", "key1", "first");
+    CHECK_REPLY_ARRAY(cc, reply, 2);
+    CHECK_REPLY_STR(cc, reply->element[0], "key1");
+    CHECK_REPLY_STR(cc, reply->element[1], "first");
+    freeReplyObject(reply);
+
+    // Multi key, but handled by same instance
+    reply = (redisReply *)redisClusterCommand(
+        cc, "eval %s 2 %s %s", "return {KEYS[1],KEYS[2]}", "key1", "key1");
+    CHECK_REPLY_ARRAY(cc, reply, 2);
+    CHECK_REPLY_STR(cc, reply->element[0], "key1");
+    CHECK_REPLY_STR(cc, reply->element[1], "key1");
+    freeReplyObject(reply);
+
+    // Error handling in EVAL
+
+    // Request without key, fails due to no key given.
+    reply = (redisReply *)redisClusterCommand(
+        cc, "eval %s 0", "return redis.call('set','foo','bar')");
+    assert(reply == NULL);
+
+    // Prepare a key to be a list-type, then run a script the attempts
+    // to access it as a simple type.
+    reply = (redisReply *)redisClusterCommand(cc, "del foo");
+    CHECK_REPLY_INT(cc, reply, 1);
+    freeReplyObject(reply);
+
+    reply = (redisReply *)redisClusterCommand(cc, "lpush foo a");
+    CHECK_REPLY_INT(cc, reply, 1);
+    freeReplyObject(reply);
+
+    reply = (redisReply *)redisClusterCommand(
+        cc, "eval %s 1 %s", "return redis.call('get',KEYS[1])", "foo");
+    CHECK_REPLY_ERROR(cc, reply, "ERR Error running script");
+    freeReplyObject(reply);
+
+    // Two keys handled by different instances,
+    // will be retried multiple times and fail due to CROSSSLOT.
+    reply = (redisReply *)redisClusterCommand(
+        cc, "eval %s 2 %s %s %s %s", "return {KEYS[1],KEYS[2],ARGV[1],ARGV[2]}",
```

Comment:
Retried multiple times??? If we get a CROSSSLOT error, there is no point retrying, is there?

Reply:
Exacly, also weird. This needs to be addressed, but the test states the current behavior. The API is not that good telling which specific error it is either, just that it failed multiple attempts. Not even that, but "too many cluster redirect"